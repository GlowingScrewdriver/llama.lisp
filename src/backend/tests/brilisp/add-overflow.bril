(brilisp
    (bril-define ((print int) (n int)))

    (bril-define ((pow int) (base int) (exp int))
        (set (out int) (const 1))
        (set (one int) (const 1))
        
        (label loop)
        (set (end bool) (lt exp one))
        (br end ret body)
        
        (label body)
        (set (out int) (mul out base))
        (set (exp int) (sub exp one))
        (jmp loop)
        
        (label ret)
        (ret out))

    (bril-define ((main int))
        (set (one int) (const 1))
        (set (two int) (const 2))
        (set (thirtyone int) (const 31))
        
        (set (half int) (call pow two thirtyone))
        (set (tmp int) (call print half))

        (ret tmp)))
