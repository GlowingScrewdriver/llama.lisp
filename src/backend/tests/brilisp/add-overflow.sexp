(brilisp
    (define ((print int) (n int)))

    (define ((pow int) (base int) (ex int))
        (set (out int) (const 1))
        (set (one int) (const 1))

        (label loop)
        (set (end bool) (lt ex one))
        (br end ret body)

        (label body)
        (set (out int) (mul out base))
        (set (ex int) (sub ex one))
        (jmp loop)

        (label ret)
        (ret out))

    (define ((main int))
        (set (one int) (const 1))
        (set (two int) (const 2))
        (set (thiry int) (const 30))

        (set (half int) (call pow two thiry))
        (set (tmp int) (call print half))

        (set (smaller int) (sub half one))
        (set (maxint int) (add half smaller))
        (set (tmp int) (call print maxint))

        (set (overflow int) (add maxint one))
        (set (tmp int) (call print overflow))
        (ret tmp)))
