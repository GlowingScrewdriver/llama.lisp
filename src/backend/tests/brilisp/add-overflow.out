1073741824
2147483647
-2147483648
